### PR TITLE
HTTP Server: Serve Grafana with a custom URL path prefix

### LINKED ISSUE #16623: Allow Grafana to be served with a URL path prefix, without a requirement for a rewriting proxy like nginx
**What would you like to be added**:
Allow Grafana to run using a real subpath, without the requirement for a mandatory path-rewriting proxy. https://github.com/grafana/grafana/issues/16613

**Why is this needed**:

* First, so that this would work - `docker run -ti --rm --name=grafana -e GF_SERVER_ROOT_URL='%(protocol)s://%(domain)s/app/grafana' -p 3000:3000 grafana/grafana`

* In some settings, like for example Ingress in Kubernetes, it is useful to not force the proxy to become a rewriting proxy but rather allow each application have its own path prefix and simply serve all its content from that path.

### PR DESCRIPTION
**Pull request by jan25**
**What this PR does / why we need it**:
Serving grafana from a non-root URL will be useful for cases of path based application routing. Currently only way to achieve this is by having a proxy with subpath rewrite rules in front of Grafana.

This PR aims to provide the subpath behavior without having to setup proxy each time we want to do path based routing.

See #16623 and #16613 for discussion around this topic.

**Which issue(s) this PR fixes**:

Fixes #16623

**Special notes for your reviewer**:
<s>Marking this PR as WIP temporarily, so reviewers could give suggestions on implementation.</s>

Currently this fix is using prefix url api from macaron, which seem to work fine on manual testing.

### CODE REVIEW

**Comment by marefr** [comment]
Thanks. This will break a lot of installs using [reverse proxy with sub path](https://grafana.com/docs/installation/behind_proxy/#examples-with-sub-path-ex-http-foo-bar-grafana) so I'm afraid you'll need to introduce a new setting to handle this. At least that's the current only solution I see right now together with documentation for us to be able to accept this feature.

**Comment by jan25** [comment]
@marefr You're right. I'll add a new setting for subpath. Should we mark the current `root_url` subpath solution as deprecated as part of this PR? Maybe we could edit docs and add a comment in `defaults.ini` file to indicate this change.
Thank you!

File: pkg/api/http_server.go [hunk 1]
```diff
@@ -227,6 +227,9 @@ func (hs *HTTPServer) addMiddlewaresAndStaticRoutes() {
 
 	m.Use(middleware.AddDefaultResponseHeaders())
 
+	if setting.AppSubUrl != "" {
+		m.SetURLPrefix(setting.AppSubUrl)
```
Comment (torkelo) [inline pkg/api/http_server.go]: nice, did not know maracon had this, thought we would have to modify all macaron routes :)

**Comment by jan25** [comment]
Added `use_proxy_setup` bool flag setting to use proxy setup or not. Defaults to true(for backward compatibility) and to not break current installs. If set to false Grafana is served from `root_url`; and no proxy required in-front.

May need to update documentation accordingly?

**Review by marefr — CHANGES_REQUESTED** [review]
Thanks. This looks quite good now. 

I still don't like the naming of the setting though and it should be defined so the value can be false per default - We try to enforce naming so that default boolean settings are false.

I don't have any great ideas of naming, but here they go anyway:
- `serve_from_sub_path = false`
- `serve_from_app_sub_url = false`
- `append_sub_path = false`
- `append_sub_path = false`

Thoughts/ideas?

Documentation needs to be updated with the new configuration parameter in this [file](https://github.com/grafana/grafana/blob/master/docs/sources/installation/configuration.md).

**Comment by jan25** [comment]
Thanks for review @marefr . `serve_from_sub_path` sounds good, I think. Also, updated the installation docs.

And, any hints whats up with codespell checker? Have we changed something in it recently? It is [failing](https://circleci.com/gh/grafana/grafana/103770?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link) for my recent commits in unrelated files.

**Review by marefr — COMMENTED** [review]
Think we can merge this soon, looks good. Still need to do some manual testing of some edge cases just to verify they works as expected. Until then, please merge your branch with latest master to resolve the codespell issues (I think).

**Review by marefr — COMMENTED** [review]
Just manually tested this and it works really nice :+1: Verified some special middleware routes:
- /metrics works as expected
- /api/health works as expected

So please merge or rebase with master and we can merge this. Thanks

**Comment by jan25** [comment]
CI passed after merging master. Thanks

**Comment by marefr** [comment]
@jan25 thanks. Pushed some configuration file and docs updates. This should be ready for merge :+1:

**Review by marefr — APPROVED** [review]

**Comment by marefr** [comment]
Thank you for contributing to Grafana!

**Comment by kesor** [comment]
Works great! Thank you!

```
docker run -ti --rm --name=grafana \
    -e GF_SERVER_ROOT_URL='%(protocol)s://%(domain)s/app/grafana' \
    -e GF_SERVER_SERVE_FROM_SUB_PATH=true \
    -p 3000:3000 \
    grafana/grafana:master
```

**Comment by Ghazgkull** [comment]
This is a great feature and I'm excited for it to be available in the next release.

One word of caution: This setting currently appears in the live configuration docs on grafana.com, although it only works when running from master. Until the feature is actually released, anyone trying to use this setting will be confused.

https://grafana.com/docs/installation/configuration/#serve-from-sub-path

**Comment by marefr** [comment]
@Ghazgkull thanks for pointing that out. We sometime miss to add that. Would you be interested to submit a fix for this? What we want to do is add an availability note like this one https://grafana.com/docs/installation/configuration/#error-or-timeout to the serve_from_sub_path setting. Let me know, otherwise we'll try to get back to this in a while and update the docs.

**Comment by Ghazgkull** [comment]
Sure, I can do that. Do you have any particular language you use for features that are not yet available in a release? Should we just assume that it will be available in the next minor version and say something like: `Will be available in v6.2.5 and above`?

**Comment by marefr** [comment]
@Ghazgkull sorry. Should be: `Available in 6.3 and above`

**Comment by ZombiePm** [comment]
http {
   ```
    map $http_upgrade $connection_upgrade {
        default upgrade;
        '' close;
    }

    upstream upgrafana {
        server 127.0.0.1:3000;
    }

    server {
       ```

        location ^~ /grafana/ {
            rewrite ^/grafana/(.*)$ /$1 break;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection $connection_upgrade;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;
            proxy_pass http://upgrafana;
        }
    }
}


docker run -ti --rm --name=grafana \
    -e GF_SERVER_ROOT_URL='https://%(domain)s/grafana' \
    -p 127.0.0.1:3000:3000 \
    grafana/grafana:master

**Pull request merged**